_posts/2025-10-24-Making-Search-Smarter-System-Generated-Search-Pipelines.markdown

```diff
@@ -0,0 +1,342 @@
+
+### Step 3: Register the factory in your plugin
+
+```java
+@Override
+public Map<String, SystemGeneratedProcessor.SystemGeneratedFactory<SearchRequestProcessor>> getSystemGeneratedRequestProcessors(
+    Parameters parameters
+) {
+    return Map.of(
+        ExampleSearchRequestPostProcessor.Factory.TYPE,
+        new ExampleSearchRequestPostProcessor.Factory()
+    );
+}
+```
+
+Once the factory is registered, OpenSearch automatically evaluates incoming search requests, generates system processors where applicable, and inserts them into the runtime search pipeline. For more examples, see the [example plugin](https://github.com/opensearch-project/OpenSearch/tree/main/plugins/examples/system-search-processor/src/main/java/org/opensearch/example/systemsearchprocessor). 
+
+Currently, OpenSearch allows only one system-generated search processor per type and stage for each search request. For example, only one system-generated search request processor can run before user-defined processors. This design simplifies execution order management and ensures predictable behavior across different plugins.
+
+In most cases, a single processor per type and stage is sufficient, but future releases may support multiple processors if use cases arise.
+
+You can also add logic in your processor to detect and handle conflicts between your system-generated processors and user-defined processors. This is useful if your processor cannot coexist with certain user-defined ones or if you need to enforce execution constraints.
+
+The following is an example of handling a conflict between a system-generated search processor and a user defined search processor:
+
+```java
+@Override
+public void evaluateConflicts(ProcessorConflictEvaluationContext context) throws IllegalArgumentException {
+    boolean hasTruncateHitsProcessor = context.getUserDefinedSearchResponseProcessors()
+        .stream()
+        .anyMatch(processor -> CONFLICT_PROCESSOR_TYPE.equals(processor.getType()));
+
+    if (hasTruncateHitsProcessor) {
+        throw new IllegalArgumentException(
+            String.format(
+                Locale.ROOT,
+                "The [%s] processor cannot be used in a search pipeline because it conflicts with the [%s] processor, "
+                    + "which is automatically generated when executing a match query against [%s].",
+                CONFLICT_PROCESSOR_TYPE,
+                TYPE,
+                TRIGGER_FIELD
+            )
+        );
+    }
+}
+```
+
+We recommend that you add a validation step to check whether a custom system-generated processor factory is enabled when a search request contains a parameter that would trigger that processor. This ensures that you receive clear error message about which factory is required, rather than having the request silently do nothing. 
```

Comment:
```suggestion
We recommend that you add a validation step to check whether a custom system-generated processor factory is enabled when a search request contains a parameter that would trigger that processor. This ensures that you receive a clear error message about which factory is required rather than having the request silently do nothing. 
```